django/core/mail/__init__.py

```diff
@@ -125,13 +130,28 @@ def mail_admins(
     """Send a message to the admins, as defined by the ADMINS setting."""
     if not settings.ADMINS:
         return
-    if not all(isinstance(a, (list, tuple)) and len(a) == 2 for a in settings.ADMINS):
-        raise ValueError("The ADMINS setting must be a list of 2-tuples.")
+    recipients = settings.ADMINS
+
+    # RemovedInDjango70Warning.
+    if all(isinstance(a, (list, tuple)) and len(a) == 2 for a in settings.ADMINS):
+        warnings.warn(
+            "Using (name, address) pairs in the ADMINS setting is deprecated."
+            " Replace with a list of email address strings.",
+            RemovedInDjango70Warning,
+        )
+        recipients = [a[1] for a in recipients]
+
+    if not isinstance(recipients, (list, tuple)) or not all(
+        isinstance(address, (str, Promise)) for address in recipients
+    ):
+        raise ImproperlyConfigured(
+            "The ADMINS setting must be a list of email address strings."
+        )
```

Comment:
Suggesting:
```suggestion
    if all(isinstance(a, (list, tuple)) and len(a) == 2 for a in settings.ADMINS):
        warnings.warn(
            "Using (name, address) pairs in the ADMINS setting is deprecated."
            " Replace with a list of email address strings.",
            RemovedInDjango70Warning,
        )
        recipients = [a[1] for a in settings.ADMINS]
    elif all(isinstance(address, (str, Promise)) for address in settings.ADMINS):
        recipients = settings.ADMINS
    else:
        raise ImproperlyConfigured(
            "The ADMINS setting must be a list of email address strings."
        )
```

- no reassignment of `recipients`, we assign once
- `if`/`elif`/`else` makes it clear exactly one branch will run
- does not accidentally fall through if someone configured a list of triples (not a big deal but)

Reply:
I was trying to isolate the portion that will be removed after deprecation.

> does not accidentally fall through if someone configured a list of triples (not a big deal but)

Neither does the original. The first commit (that updates the tests) adds a test for that specific case.

Also, I think this suggestion incorrectly handles `ADMINS = "single-email-address@example.com"` by emailing to `["s", "i", "n", "g", ...]`.

Reply:
Iterating over strings giving strings is the curse of our time.... anyways, given that you unified these code paths anyways I'm not that worried about this anymore. Please ignore